pyro/contrib/forecast/util.py

```diff
@@ -0,0 +1,231 @@
+        assert model_value.dim() == guide_value.dim()
+        for dim in range(model_value.dim()):
+            if model_value.size(dim) != guide_value.size(dim):
+                break
+        assert model_value.size(dim) > guide_value.size(dim)
+        assert model_value.shape[dim + 1:] == guide_value.shape[dim + 1:]
+        split = guide_value.size(dim)
+        index = (slice(None),) * dim + (slice(split, None),)
+        msg["value"] = torch.cat([guide_value, model_value[index]], dim=dim)
+
+
+class PrefixConditionMessenger(Messenger):
+    """
+    EXPERIMENTAL Given a prefix of t-many observations, condition a (t+f)-long
+    distribution on the observations, converting it to an f-long distribution.
+
+    :param dict data: A dict mapping site name to tensors of observations.
+    """
+    def __init__(self, data):
+        super().__init__()
+        self.data = data
+
+    def _pyro_sample(self, msg):
+        if msg["name"] not in self.data:
+            return
+
+        assert msg["value"] is None
+        data = self.data[msg["name"]]
+        msg["fn"] = prefix_condition(msg["fn"], data)
+
+
+# ----------------------------------------------------------------------------
+# The pattern-matching code in the remainder of this file could be eventually
+# replace by much simpler Funsor logic.
+
+UNIVARIATE_DISTS = {
+    dist.Cauchy: ("loc", "scale"),
+    dist.Laplace: ("loc", "scale"),
+    dist.Normal: ("loc", "scale"),
+    dist.Stable: ("stability", "skew", "scale", "loc"),
+    dist.StudentT: ("df", "loc", "scale"),
+}
+
+
+@singledispatch
+def prefix_condition(d, data):
+    """
+    EXPERIMENTAL Given a distribution ``d`` of shape ``batch_shape + (t+f, d)``
+    and data ``x`` of shape ``batch_shape + (t, d)``, find a conditional
+    distribution of shape ``batch_shape + (f, d)``. Typically ``t`` is the
+    number of training time steps, ``f`` is the number of forecast time steps,
+    and ``d`` is the data dimension.
+
+    :param d: a distribution with ``len(d.shape()) >= 2``
+    :type d: ~pyro.distributions.Distribution
+    :param data: data of dimension at least 2.
+    :type data: ~torch.Tensor
+    """
+    try:
+        return d.prefix_condition(data)
+    except AttributeError:
+        raise NotImplementedError("prefix_condition() does not suport {}".format(type(d)))
+
+
+@prefix_condition.register(dist.Independent)
+def _(d, data):
+    base_dist = prefix_condition(d.base_dist, data)
+    return base_dist.to_event(d.reinterpreted_batch_ndims)
+
+
+def _prefix_condition_univariate(d, data):
+    t = data.size(-2)
+    params = [getattr(d, name)[..., t:, :]
+              for name in UNIVARIATE_DISTS[type(d)]]
+    return type(d)(*params)
+
+
+for _type in UNIVARIATE_DISTS:
+    prefix_condition.register(_type)(_prefix_condition_univariate)
+
+
+@prefix_condition.register(dist.MultivariateNormal)
+def _(d, data):
+    t = data.size(-2)
+    loc = d.loc[..., t:, :]
+    scale_tril = d.scale_tril[..., t:, :, :]
```

Comment:
this sort of thing only works because everything has been (possibly) expanded by now right?

Reply:
That's right, all inputs are expanded via `broadcast_all()`. Note that we could be more efficient using Du's `._unbroadcasted_scale_tril` but that optimization seems premature at this time.